trust-quorum/src/node.rs

```diff
@@ -196,21 +210,133 @@ impl Node {
+                );
+                return Ok(None);
+            }
+
+            // Idempotent request
+            if msg.config == latest_prepare.config {
+                return Ok(None);
+            }
+
+            // Does the last committed epoch match what we have?
+            let msg_last_committed_epoch =
+                msg.config.previous_configuration.as_ref().map(|p| p.epoch);
+            let last_committed_epoch =
+                self.persistent_state.last_committed_epoch();
+            if msg_last_committed_epoch != last_committed_epoch {
+                // If the msg contains an older last_committed_epoch than what
+                // we have, then out of order commits have occurred, as we know
+                // this prepare is later than what we've seen. This is a critical
+                // protocol invariant that has been violated.
+                //
+                // If the msg contains a newer last_committed_epoch than what
+                // we have, then we have likely missed a commit and are behind
+                // by more than one reconfiguration. The protocol currently does
+                // not allow this. Future protocol implementations may provide a
+                // capability to "jump" configurations.
+                //
+                // If there is a `None`/`Some` mismatch, then again, an invariant
+                // has been violated somewhere. The coordinator should know
+                // whether this is a "new" node or not, which would have a "None"
+                // configuration.
+                let alarm = Alarm::PrepareLastCommittedEpochMismatch {
+                    from,
+                    prepare_last_committed_epoch: msg_last_committed_epoch,
+                    persisted_prepare_last_committed_epoch:
+                        last_committed_epoch,
+                };
+                error!(self.log, "{alarm}");
+                return Err(alarm);
+            }
+
+            // The prepare is up-to-date with respect to our persistent state
+        };
+
+        // Are we currently trying to coordinate?
+        if let Some(cs) = &self.coordinator_state {
+            let coordinating_epoch = cs.reconfigure_msg().epoch();
+            if coordinating_epoch > msg.config.epoch {
+                warn!(self.log, "Received stale prepare while coordinating";
+                    "from" => %from,
+                    "msg_epoch" => %msg.config.epoch,
+                    "epoch" => %cs.reconfigure_msg().epoch()
+                );
+                return Ok(None);
+            }
+            if coordinating_epoch == msg.config.epoch {
+                let alarm = Alarm::DifferentNodesCoordinatingSameEpoch {
+                    epoch: coordinating_epoch,
+                    them: from,
+                    us: self.platform_id.clone(),
+                };
+                error!(self.log, "{alarm}");
+                return Err(alarm);
+            }
```

Comment:
apologies, having trouble following this too. again, it seems like there are 3-4 epochs at play here, and it's not clear to me what the relationship is among all of them. Maybe just an ASCII diagram at the top of the file/function would help.